### PR TITLE
Feat/graph ql transformer core added bracket mismatch detection

### PR DESCRIPTION
**Pull request by ncarvajalc**
#### Description of changes

Added bracket validation for graphQL schema.
Added custom error for bracket mismatch.

This was necessary as the `parse` function from the graphql dependency doesn't tell much about the error and it isn't clear it is because of a bracket mismatch.

Error without the changes: 
```bash
🛑 Syntax Error: Expected Name, found {
``` 

Error after the changes:
```bash
🛑 Syntax Error: mismatched brackets found in the schema. Unexpected )
```
or
```bash
🛑 Syntax Error: mismatched brackets found in the schema. Missing )
```

#### Issue #, if available
#107

#### Description of how you validated changes

`yarn test` passing and used custom app with `amplify-dev` to see the newly thrown errors when there is a bracket mismatch.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-api/src/graphql-transformer/sandbox-mode-helpers.ts [hunk 1]
```diff
@@ -37,10 +42,38 @@ function matchesGlobalAuth(field: any): boolean {
...
+      case '[':
+        stack.push(']');
+        break;
+      case '{':
+        stack.push('}');
+        break;
+      case ')':
+      case ']':
+      case '}':
+        if (c !== stack.pop()) {
```
Comment (phani-srikar) [inline packages/amplify-category-api/src/graphql-transformer/sandbox-mode-helpers.ts:62]: I like how you used the fallthrough case statements.

File: packages/amplify-category-api/src/graphql-transformer/sandbox-mode-helpers.ts [hunk 2]
```diff
@@ -37,10 +42,38 @@ function matchesGlobalAuth(field: any): boolean {
...
+    switch (c) {
+      case '(':
+        stack.push(')');
+        break;
+      case '[':
+        stack.push(']');
+        break;
+      case '{':
+        stack.push('}');
+        break;
```
Comment (phani-srikar) [inline packages/amplify-category-api/src/graphql-transformer/sandbox-mode-helpers.ts:58]: Can we use similar fallthrough like you used below to say `stack.push(c)` once?

**Review by phani-srikar — CHANGES_REQUESTED** [review]
Can you please add a few test cases covering some edge cases like:
- Empty schema i.e '' is handled.
- Schema with mismatched brackets throws appropriate error.
- Schema with correct brackets does not throw.
Thanks

**Comment by ncarvajalc** [comment]
@phani-srikar with @jcollinc we made the changes you requested. Hope it gets merged soon :D

**Review by phani-srikar — APPROVED** [review]
LGTM. Thanks!

**Comment by alharris-at** [comment]
Hey @ncarvajalc - I see a bunch of `yarn.lock` updates that don't seem related to this change. (Probably related to running `setup-dev`. Could you `rm` that file and just run `yarn` in the root directory to rebuild?

**Review by alharris-at — COMMENTED** [review]
LGTM - just left one comment about the `yarn.lock` update.

**Comment by ncarvajalc** [comment]
@alharris-at with @jcollinc we added the line where the bracket mismatch is occurring. Let me know what you think.

**Review by alharris-at — APPROVED** [review]

**Review by alharris-at — CHANGES_REQUESTED** [review]
Discussed offline, as implemented this will currently throw an error on a comment which has a mismatch brace in it, which constitutes a breaking change for customers.

Previously this line would have been part of a legal schema, but now would now.

```graphql
# This is my type that has some parameters :)
```

Recommend switching the response to a warning instead of an exception.

**Comment by ncarvajalc** [comment]
@alharris-at with @jcollinc we successfully solved the issue we had and now it doesn't consider comments or strings for the bracket check, as there can be cognito group names with hashes and brackets in it. Now this is valid:
`# This is my type that has some parameters :)`
And also this:
```graphql
@auth(
    rules: [
      {
        allow: owner
        groups: ["Admins#(accounts):)"]
        operations: [create, update, delete]
      }
      { allow: public, operations: [read] }
    ]
  ) 
```

**Review by alharris-at — APPROVED** [review]

**Review by AaronZyLee — COMMENTED** [review]
The change overall LGTM. Left some comments below.

- By removing `yarn.lock` it means removing them in the git commit. Sorry for the confusion.

**Comment by alharris-at** [comment]
Closing in favor of https://github.com/aws-amplify/amplify-category-api/pull/1020
